### DISCUSSION TITLE
Conditionally applying aspect

### DISCUSSION THREAD

**Discussion by lukedays** [metalama/Metalama#30] [Q&A] [answered]
Hi all,

First of all, thanks a lot for the library!
Is it possible to conditionally apply the OverrideMethodAspect based on parameters?

Example: 

```csharp
if (!meta.Target.Parameters["cache"].Value)
{
    return meta.Proceed(); // Don't cache
}
else
{
    // Cache logic...
}
```

This gives an error: Cannot implicitly convert type 'object' to 'bool'.
I tried casting to bool with no success.

Thanks in advance!

**Comment by gfraiteur** [comment, marked as answer]
This code (open in [online sandbox](https://try.metalama.net/#N4IghiBcIPQwBAFQBYEsDO8PzAO3gPYAOAprgLToECuATgMYnwCyJALmADZgC2Y8JAB68inEgDp4AZRJNkbNkXSQ4Ac1Rtk1AEbj6BHjCIF0bdMjC0iMVh259xUkWMwAzArXg8PEgDq5-anRUXFUWdi5eMHEAMVpeEgB3DwBrcQBBdFJ6MwBuQODQ8Lso2PieJNTxAGECABMSfNwgkLDbSIc4hOTaNNqGxwBPXA5BACFqVE4G2nQmlqKpQdMSHhqCTjEc1AJcdHXcejpaMjYm-yIdTlR6eHpudExqsHpkEnSFWlRtajYmSHgAHkAG4kWhfBq2ZD1TLZNj+YD+eDI+CXbTXW4EUHg1ANeB1Ya8G4AfiB2IhJCh9QAFABKJEoxH4FEsuDwCZTOqYTRMegvNBFUxfULiBks5HAyx3fkkADSJEG8AAvPAAOLsZ6vVrywYxDx8BStKRsYWqOniRAEABqXGoJDp52Z4vgks8fNeJAACpYEn9PCqKhwLZZVOxxN7yuwwfspK0xIDaAAREiuMDUThsamo5UAPlR4gAcgllSrfCB0ClUERNW8y-BaY7ncjUK54Fn3W8I76wfAAIQq5qbeti8VMpviltttvaAgbWnSj1dwNg8Q2zh24dO8fwMfb50nNh0fCB6Ke2gERgkOoOkfjgC+t5RD4CW9ZCBrTE4s5S1CIotfKKTlmTg8KIJAfjUMoWrQgzqmwa52u2Mo6gANIQvyENoABWJA5C6tpMPO9IAciu57myH7wGgbD-nuyIHke+Hro0j7Is+44kJw6AkKxO68W+8CUTwGD7GqBA4LgdRYLgwIECkTA8l47DQnUtF0a68AnOg6ZsMqSlBmeF6yNeDb+PxKJsukdRSWw4kdn4JHiiBYEQRBiAwVZ14Lm8qGaSQ2kZvWjZ0QxtD4FpOlNPeI7PiORBfJKfzwKYYBsDc8AAJIjGCxjcH8dTGqaHLTD2cE1tqCp6rQBppaEhWtHSI5keKGlCq0xUzHpuBJJl2VWBsqVXvVoQdWCN6OW1I2TCVtAZNZiBCJm+nRIgIZhoggykBaBDDWaRFRU2k2qKNs2eQtghLWWoogEFvFHSdc11OdS0nsGtChjRVKqUWFS3RNJrtdNMyPc9WZltSdamS+44aagemvbUoFTCQiCoBUWYAAx-bx7gnPyU4aUQ0nLW9H3hj6y6zJu47NTDUr6DwfB6XDeZY6SZZoXWAJlmWwV7kB+YAEoprKIRSUqKrC64ouSeIgIYcRdF8Y5473UDK5nYtWYACRlsADN8HewB-j9JB3npAA8NBsHmkMHXu7F0Zx3HmaRrvimrnIa-NWvwLrID6wYhvAIiN1Q0rLKezNj0AKKCPF-nBLsWZ-ghhH2xHUfA5rF1gyAD5hxn0XQ3RqAANRl0XT5mf9RXq6dPu5-AZb0oXNfjqF+BZ2CGfPrFuBsp63CMNCM1YKBYgVCMqU7PgBCtvw9nwNouFptx8CKdCKw2TB4gnvY0TdbpdQEP58C4AQulcF+iTwLj4TeDBEH+CEfq4FwyUcGltz3GAjzSM4cCMoERxSuOlFK38-JgBPrgTgipaiHGOKcRMNw0q7EsIMC2R00IEGwrhG2gkZRdSSONO8IAUIgCIFAWACAUDYGwHgQgpAKBUDoIwYoHR+BCEAZIGQcgFBKBUDAdQmgdB6AMEYEwZgLBWBsBEA+jhAFuA8F4Hw-4X59XfpwO4DxMCGVUOUEBW54qoESkwCB6VZK4hYGAEIjUty0xRAgqgYhxAAHUvh-AADIhHtM3EAzxNitBwHfVAsxdJpQqNdP645PJZgAIxoXiTEpsziNgSA8RoEgPjup50CdcIo-BuL6EkhvNGfg26OTifARJNSUnOjSa4zJ3jfF5IMGBfK0Tw7VxLsiAA2h+AAunFBKg1P4z1uK-eA1SplgDQlM7QaEZwbGSpWasRCVSpi4oRJqvFGkZM8dk1p-i6G4ErCKcQXS+bik7iEsuy9e7+DIXeIAA)) works for me:

```cs
    public override dynamic? OverrideMethod()
    {
        // Builds the caching string.
        var cacheKey = GetCachingKeyFormattingString().ToValue();

        var cacheParameter = meta.Target.Parameters.SingleOrDefault( p => p.Name == "skipCache" );

        if ( cacheParameter != null )
        {
            if ( (bool) cacheParameter.Value )
            {
                return meta.Proceed();
            }
        }

        // Cache lookup.
        if ( SampleCache.Cache.TryGetValue( cacheKey, out object value ) )
        {
            // Cache hit.
            return value;
        }
        else
        {
            // Cache miss. Go and invoke the method.
            var result = meta.Proceed();

            // Add to cache.
            SampleCache.Cache.TryAdd( cacheKey, result );
      }
   }
```

**Reply by lukedays** [reply]
Hello, thanks for looking into this!
Unfortunately it didn't work. The function I'm applying the aspect has a generic parameter, and the example stops working as well if I add one.
I managed to reproduce the error on my machine ([repo](https://github.com/lukedays/TestMetalamaParameters)) and [on the sandbox](https://try.metalama.net/#N4IghiBcIPQwBAFQBYEsDO8PzAO3gPYAOAprgLToECuATgMYnwCyJALmADZgC2Y8JAB68inEgDp4AZRJNkbNkXSQ4Ac1Rtk1AEbj6BHjCIF0bdMjC0iMVh259xUkWMwAzArXg8PEgDq5-anRUXFUWdi5eMHEAMVpeEgB3DwBrcQBBdFJ6MwBuQODQ8Lso2PieJNTxAGECABMSfNwgkLDbSIc4hOTaNNqGxwBPXA5BACFqVE4G2nQmlqKpQdMSHhqCTjEc1AJcdHXcejpaMjYm-yIdTlR6eHpudExqsHpkEnSFWlRtajYmSHgAHkAG4kWhfBq2ZD1TLZNj+YD+eDI+CXbTXW4EUHg1ANeB1Ya8G4AfiB2IhJCh9QAFABKJEoxH4FEsuDwCZTOqYTRMegvNBFUxfULiBks5HAyx3fkkADSJEG8AAvPAAOLsZ6vVrywYxDx8BStKRsYWqOniRAEABqXGoJDp52Z4vgks8fNeJAACpYEn9PCqKhwLZZVOxxN7yuwwfspK0xIDaAAREiuMDUThsamo5UAPlR4gAcgllSrfCB0ClUERNW8y-BaY7ncjUK54Fn3W8I76wfAAIQq5qbeti8VMpviltttvaAgbWnSj1dwNg8Q2zh24dO8fwMfb50nNh0fCB6Ke2gERgkOoOkfjgC+t5RD4CW9ZCBrTE4s5S1CIotfKKTlmTg8KIJAfjUMoWrQgzqmwa52u2Mo6gANIQvyENoABWJA5C6tpMPO9IAciu57myH7wGgbD-nuyIHke+Hro0j7Is+44kJw6AkKxO68W+8CUTwGD7GqBA4LgdRYLgwIECkTA8l47DQnUtF0a68AnOg6ZsMqSlBmeF6yNeDb+PxKJsukdRSWw4kdn4JHiiBYEQRBiAwVZ14Lm8qGaSQ2kZvWjZ0QxtD4FpOlNPeI7PiORBfJKfzwKYYBsDc8AAJIjGCxjcH8dTGqaHLTD2cE1tqCp6rQBppaEhWtHSI5keKGlCq0xUzHpuBJJl2VWBsqVXvVoQdWCN6OW1I2TCVtAZNZiBCJm+nRIgIZhoggykBaBDDWaRFRU2k2qKNs2eQtghLWWoogEFvFHSdc11OdS0nsGtChjRVKqUWFS3RNJrtdNMyPc9WZltSdamS+44aagemvbUoFTCQiCoBUWYAAx-bx7gnPyU4aUQ0nLW9H3hj6y6zJu47NTDUr6DwfB6XDeZY6SZZoXWAJlmWwV7kB+YAEoprKIRSUqKrC64ouSeIgIYcRdF8Y5473UDK5nYtWYACRlsADN8HewB-j9JB3npAA8NBsHmkMHXu7F0Zx3HmaRrvimrnIa-NWvwLrID6wYhvAIiN1Q0rLKezNj0AKKCPF-nBLsWZ-ghhH2xHUfA5rF1gyAD5hxn0XQ3RqAANRl0XT5mf9RXq6dPu5-AZb0oXNfjqF+BZ2CGfPrFuBsp63CMNCM1YKBYgVCMqU7PgBCtvw9nwNouFptx8CKdCKw2TB4gnvY0TdbpdQEP58C4AQulcF+iTwLj4TeDBEH+CEfq4FwyUcGltz3GAjzSM4cCMoERxSuOlFK38-JgBPrgTgipaiHGOKcRMNw0q7EsIMC2R00IEGwrhG2gkZRdSSONO8IAUIgCIFAEA78KhZBeEwBaph2gHyXFGWYjoLhgJ-g8TAhlVDlBAVuNEGJP4z1uLJXELAwAhEaluWmKIEFUDEOIAA6l8P4AAZEI9oyzPE2K0HAd9UCzF0mlCo11w5Nk8hbdU3Uvj0A2qQHM1IACMaFXFWOdEojYEh1EaBINo7qEMQD6OuEUfg3F9CSQ3mjPwIAvHihsXYsENwnEkBce4+Aniq7Ih8So-xWidEhMRmBfKlje41wANofgALqgPROAr+6VX7wBsYgFxrSwBoVadoNCM4NjJUrNWIhKpUxcRIIrRkvF8l+I0YE4pZYUAhErCKcQFT25Nk7kYsuy9Km4H7iI9Kv9-4pIceknc8Bnz5yAA).


Program.cs(21,28): error CS0117: 'Program' does not contain a definition for 'Add_Source'

Thanks in advance!

**Reply by gfraiteur** [reply]
I have created a [new issue](https://github.com/postsharp/Metalama/issues/32) for this. Please follow the new issue to get pinged.
